### ISSUE TITLE
never patterns: `!` argument not detected as diverging on async fn

### ISSUE THREAD

**Issue by Nadrieril** [rust-lang/rust#120240]
With https://github.com/rust-lang/rust/pull/120104, a function with a `!` pattern in a binding is detected as diverging:
```rust
fn never_arg(!: Void) -> ! {}
fn never_let(x: Void) -> ! {
  let ! = x;
}
```

This doesn't work the arguments of an `async` fn however:
```rust
async fn never_let(x: Void) -> ! {
  let ! = x;
} // ok
async fn never_arg(!: Void) -> ! {} // ERROR mistmatched types
```

**Comment by traviscross** [comment]
@rustbot labels +AsyncAwait-Triaged +WG-async

We reviewed this today in WG-async triage.  We agree this is a bug.

**Comment by compiler-errors** [comment]
I looked into this and it has to do with the fact that we desugar the latter (`never_arg`) into something like:

```
fn never_arg(x: Void) -> ! {
    async move {
        let ! = x;
        let body = {
        };
        body
    }
}
```

I don't believe this is going to be easy or perhaps even possible to fix given the way that we don't/can't(?) make the inner `body` block (which corresponds to the body of the original async fn) to be diverging based off of a pattern in the *outer* block.

**Comment by Nadrieril** [comment]
And I presume the desugaring is this way because of drop order subtleties?

**Comment by compiler-errors** [comment]
yep
